spec/index.bs

```diff
@@ -8086,21 +8093,62 @@ interface GPUPresentationContext {
     HTML processing model, each {{GPUPresentationContext}} |context| must <dfn abstract-op>present the
     presentation context content to the canvas</dfn> by running the following steps:
 
-        1. Let |texture| be |context|.{{GPUPresentationContext/[[currentTexture]]}}.
-        1. If |texture| is `null`, stop.
-        1. Set |context|.{{GPUPresentationContext/[[currentTexture]]}} to `null`.
-        1. Ensure that all submitted work items (e.g. queue submissions) has
-            completed writing to |texture|.
-        1. Update |context|.{{GPUPresentationContext/[[canvas]]}} with the
-            contents of |texture|.
-        1. Call {{GPUTexture/destroy()}} on |texture|.
-
-    Issue: The texture should mark its `[[destroyed]]` field as true rather than calling the
-    {{GPUTexture/destroy()}} method if we separate object invalid and destroyed states.
+    1. If |context|.{{GPUPresentationContext/[[currentTexture]]}} is not `null` and
+        |context|.{{GPUPresentationContext/[[currentTexture]]}}.{{GPUTexture/[[destroyed]]}} is `false`:
+        1. Let |imageContents| be [$get the image contents of a presentation context|the image contents$] of |context|.
+        1. Update |context|.{{GPUPresentationContext/[[canvas]]}} with |imageContents|.
+        1. Call |context|.{{GPUPresentationContext/[[currentTexture]]}}.{{GPUTexture/destroy()}}.
```

Comment:
Not a blocker, but I feel like there's a slight discrepancy between this algorithm that updates the canvas with `imageContents` and the description of `[[destroyed]]` which says "If the texture is destroyed... its underlying memory can be freed." I would assume that the `imageContents` are part of the texture's "underlying memory", and as such depending on how the canvas handles the update (which is implementation specific, IIRC) this may imply that the image could be freed while in use by the canvas.

I don't have a suggestion for how to resolve that, and I don't think you should hold off merging this over it. Maybe worth an inline issue, though?

Reply:
The idea here is that `imageContents` is a copy of the image data. Implementations can optimize away the copy under the as-if principle. What I should do here is make "get the image contents" say "get a copy of the image contents", just like "get a copy of the bytes held by the buffer source":
https://heycam.github.io/webidl/#dfn-get-buffer-source-copy